rclpy/rclpy/node.py

```diff
@@ -248,14 +329,29 @@ def set_parameters(self, parameter_list: List[Parameter]) -> List[SetParametersR
         For each successfully set parameter, a :class:`ParameterEvent` message is
         published.
 
+        If setting a parameter fails due to not being declared, then the
+        parameters which have already been set will stay set, and no attempt will
+        be made to set the parameters which come after.
+
         :param parameter_list: The list of parameters to set.
         :return: A list of SetParametersResult messages.
         """
+        return self._set_parameters(parameter_list, self.set_parameters_atomically)
```

Comment:
I was confused for a minute. Maybe it should be clarified that the `setter_func` parameter is applied individually for each parameter. Otherwise, this looks like we might be setting the parameter list atomically, contrary to the docstring.

Reply:
Good idea.
I still need to work a bit on the documentation; it still needs a second pass.
In general terms I'm reusing the existing functionality as much as possible. Following that line, these are some ideas behind this approach:
- The setter methods will be split in two: a "public" part and a "private" part (with trailing underscore).
- "Private" setters won't check whether the parameters were declared or not, so they will be used to actually declare the parameters.
- "Public" setters will comply with the new logic: raise exceptions if undeclared parameters are not allowed, or implicitly declare and set the parameters if undeclared parameters are allowed.